# Cargar los datos

In [1]:
import pandas as pd, numpy as np

In [2]:
url = "https://raw.githubusercontent.com/z3r0st/EB-2022-1-CC50/main/data/CAvideos_cc50.csv"
df = pd.read_csv(url)

In [3]:
url = "https://raw.githubusercontent.com/z3r0st/EB-2022-1-CC50/main/data/CA_category_id.json"
categories = pd.read_json(url)

# Inspección de los datos

## DataFrame Principal

In [4]:
df.head()

video_id trending_date  \
0  n1WpP7iowLc      17.14.11   
1  0dBIkQ4Mz1M      17.14.11   
2  5qpjK5DgCt4      17.14.11   
3  d380meD0W0M      17.14.11   
4  2Vv-BfVoq4g      17.14.11   

                                               title channel_title  \
0         Eminem - Walk On Water (Audio) ft. Beyoncé    EminemVEVO   
1                      PLUSH - Bad Unboxing Fan Mail     iDubbbzTV   
2  Racist Superman | Rudy Mancuso, King Bach & Le...  Rudy Mancuso   
3                           I Dare You: GOING BALD!?      nigahiga   
4        Ed Sheeran - Perfect (Official Music Video)    Ed Sheeran   

   category_id              publish_time  \
0           10  2017-11-10T17:00:03.000Z   
1           23  2017-11-13T17:00:00.000Z   
2           23  2017-11-12T19:05:24.000Z   
3           24  2017-11-12T18:01:41.000Z   
4           10  2017-11-09T11:04:14.000Z   

                                                tags     views    likes  \
0  Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...  17158579   787425   
1  plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   1014651   127794   
2  racist superman|"rudy"|"mancuso"|"king"|"bach"...   3191434   146035   
3  ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...   2095828   132239   
4  edsheeran|"ed sheeran"|"acoustic"|"live"|"cove...  33523622  1634130   

   dislikes  comment_count                                  thumbnail_link  \
0     43420         125882  https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
1      1688          13030  https://i.ytimg.com/vi/0dBIkQ4Mz1M/default.jpg   
2      5339           8181  https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg   
3      1989          17518  https://i.ytimg.com/vi/d380meD0W0M/default.jpg   
4     21082          85067  https://i.ytimg.com/vi/2Vv-BfVoq4g/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   
2              False             False                   False   
3              False             False                   False   
4              False             False                   False   

                                         description             state  \
0  Eminem's new track Walk on Water ft. Beyoncé i...          Manitoba   
1  STill got a lot of packages. Probably will las...  British Columbia   
2  WATCH MY PREVIOUS VIDEO ▶ \n\nSUBSCRIBE ► http...     New Brunswick   
3  I know it's been a while since we did this sho...      Saskatchewan   
4  🎧: https://ad.gt/yt-perfect\n💰: https://atlant...           Alberta   

         lat         lon                                geometry  
0  50.150025  -96.883322        POINT (-96.88332178 50.15002545)  
1  49.099960 -116.516697         POINT (-116.516697 49.09996035)  
2  45.267042  -66.076675  POINT (-66.07667505000001 45.26704185)  
3  50.933311 -102.799989        POINT (-102.7999891 50.93331097)  
4  53.016698 -112.816639        POINT (-112.8166386 53.01669802)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40881 entries, 0 to 40880
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   video_id                40881 non-null  object 
 1   trending_date           40881 non-null  object 
 2   title                   40881 non-null  object 
 3   channel_title           40881 non-null  object 
 4   category_id             40881 non-null  int64  
 5   publish_time            40881 non-null  object 
 6   tags                    40881 non-null  object 
 7   views                   40881 non-null  int64  
 8   likes                   40881 non-null  int64  
 9   dislikes                40881 non-null  int64  
 10  comment_count           40881 non-null  int64  
 11  thumbnail_link          40881 non-null  object 
 12  comments_disabled       40881 non-null  bool   
 13  ratings_disabled        40881 non-null  bool   
 14  video_error_or_removed  40881 non-null

In [6]:
df.isna().sum()

video_id                     0
trending_date                0
title                        0
channel_title                0
category_id                  0
publish_time                 0
tags                         0
views                        0
likes                        0
dislikes                     0
comment_count                0
thumbnail_link               0
comments_disabled            0
ratings_disabled             0
video_error_or_removed       0
description               1296
state                        0
lat                          0
lon                          0
geometry                     0
dtype: int64

## Categorías

In [7]:
categories.head()

kind  \
0  youtube#videoCategoryListResponse   
1  youtube#videoCategoryListResponse   
2  youtube#videoCategoryListResponse   
3  youtube#videoCategoryListResponse   
4  youtube#videoCategoryListResponse   

                                                etag  \
0  "ld9biNPKjAjgjV7EZ4EKeEGrhao/1v2mrzYSYG6onNLt2...   
1  "ld9biNPKjAjgjV7EZ4EKeEGrhao/1v2mrzYSYG6onNLt2...   
2  "ld9biNPKjAjgjV7EZ4EKeEGrhao/1v2mrzYSYG6onNLt2...   
3  "ld9biNPKjAjgjV7EZ4EKeEGrhao/1v2mrzYSYG6onNLt2...   
4  "ld9biNPKjAjgjV7EZ4EKeEGrhao/1v2mrzYSYG6onNLt2...   

                                               items  
0  {'kind': 'youtube#videoCategory', 'etag': '"ld...  
1  {'kind': 'youtube#videoCategory', 'etag': '"ld...  
2  {'kind': 'youtube#videoCategory', 'etag': '"ld...  
3  {'kind': 'youtube#videoCategory', 'etag': '"ld...  
4  {'kind': 'youtube#videoCategory', 'etag': '"ld...

In [8]:
categories.index

RangeIndex(start=0, stop=31, step=1)

In [9]:
categories.columns

Index(['kind', 'etag', 'items'], dtype='object')

In [10]:
categories["items"].head()

0    {'kind': 'youtube#videoCategory', 'etag': '"ld...
1    {'kind': 'youtube#videoCategory', 'etag': '"ld...
2    {'kind': 'youtube#videoCategory', 'etag': '"ld...
3    {'kind': 'youtube#videoCategory', 'etag': '"ld...
4    {'kind': 'youtube#videoCategory', 'etag': '"ld...
Name: items, dtype: object

In [11]:
categories.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31 entries, 0 to 30
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   kind    31 non-null     object
 1   etag    31 non-null     object
 2   items   31 non-null     object
dtypes: object(3)
memory usage: 872.0+ bytes


In [12]:
categories["items"].values

array([{'kind': 'youtube#videoCategory', 'etag': '"ld9biNPKjAjgjV7EZ4EKeEGrhao/Xy1mB4_yLrHy_BmKmPBggty2mZQ"', 'id': '1', 'snippet': {'channelId': 'UCBR8-60-B28hp2BmDPdntcQ', 'title': 'Film & Animation', 'assignable': True}},
       {'kind': 'youtube#videoCategory', 'etag': '"ld9biNPKjAjgjV7EZ4EKeEGrhao/UZ1oLIIz2dxIhO45ZTFR3a3NyTA"', 'id': '2', 'snippet': {'channelId': 'UCBR8-60-B28hp2BmDPdntcQ', 'title': 'Autos & Vehicles', 'assignable': True}},
       {'kind': 'youtube#videoCategory', 'etag': '"ld9biNPKjAjgjV7EZ4EKeEGrhao/nqRIq97-xe5XRZTxbknKFVe5Lmg"', 'id': '10', 'snippet': {'channelId': 'UCBR8-60-B28hp2BmDPdntcQ', 'title': 'Music', 'assignable': True}},
       {'kind': 'youtube#videoCategory', 'etag': '"ld9biNPKjAjgjV7EZ4EKeEGrhao/HwXKamM1Q20q9BN-oBJavSGkfDI"', 'id': '15', 'snippet': {'channelId': 'UCBR8-60-B28hp2BmDPdntcQ', 'title': 'Pets & Animals', 'assignable': True}},
       {'kind': 'youtube#videoCategory', 'etag': '"ld9biNPKjAjgjV7EZ4EKeEGrhao/9GQMSRjrZdHeb1OEM1XVQ9zbGec"', '

In [13]:
categories["items"].values.tolist()

[{'kind': 'youtube#videoCategory',
  'etag': '"ld9biNPKjAjgjV7EZ4EKeEGrhao/Xy1mB4_yLrHy_BmKmPBggty2mZQ"',
  'id': '1',
  'snippet': {'channelId': 'UCBR8-60-B28hp2BmDPdntcQ',
   'title': 'Film & Animation',
   'assignable': True}},
 {'kind': 'youtube#videoCategory',
  'etag': '"ld9biNPKjAjgjV7EZ4EKeEGrhao/UZ1oLIIz2dxIhO45ZTFR3a3NyTA"',
  'id': '2',
  'snippet': {'channelId': 'UCBR8-60-B28hp2BmDPdntcQ',
   'title': 'Autos & Vehicles',
   'assignable': True}},
 {'kind': 'youtube#videoCategory',
  'etag': '"ld9biNPKjAjgjV7EZ4EKeEGrhao/nqRIq97-xe5XRZTxbknKFVe5Lmg"',
  'id': '10',
  'snippet': {'channelId': 'UCBR8-60-B28hp2BmDPdntcQ',
   'title': 'Music',
   'assignable': True}},
 {'kind': 'youtube#videoCategory',
  'etag': '"ld9biNPKjAjgjV7EZ4EKeEGrhao/HwXKamM1Q20q9BN-oBJavSGkfDI"',
  'id': '15',
  'snippet': {'channelId': 'UCBR8-60-B28hp2BmDPdntcQ',
   'title': 'Pets & Animals',
   'assignable': True}},
 {'kind': 'youtube#videoCategory',
  'etag': '"ld9biNPKjAjgjV7EZ4EKeEGrhao/9GQMSRjrZdHe

### Visualizar valores de la columna items

In [14]:
import urllib.request, json

In [15]:
url = "https://raw.githubusercontent.com/z3r0st/EB-2022-1-CC50/main/data/CA_category_id.json"                   
with urllib.request.urlopen(url) as file:
    cat_data = json.loads(file.read().decode())
items = pd.json_normalize(cat_data, "items", record_prefix="item_")

In [16]:
items.head()

item_kind                                          item_etag  \
0  youtube#videoCategory  "ld9biNPKjAjgjV7EZ4EKeEGrhao/Xy1mB4_yLrHy_BmKm...   
1  youtube#videoCategory  "ld9biNPKjAjgjV7EZ4EKeEGrhao/UZ1oLIIz2dxIhO45Z...   
2  youtube#videoCategory  "ld9biNPKjAjgjV7EZ4EKeEGrhao/nqRIq97-xe5XRZTxb...   
3  youtube#videoCategory  "ld9biNPKjAjgjV7EZ4EKeEGrhao/HwXKamM1Q20q9BN-o...   
4  youtube#videoCategory  "ld9biNPKjAjgjV7EZ4EKeEGrhao/9GQMSRjrZdHeb1OEM...   

  item_id    item_snippet.channelId item_snippet.title  \
0       1  UCBR8-60-B28hp2BmDPdntcQ   Film & Animation   
1       2  UCBR8-60-B28hp2BmDPdntcQ   Autos & Vehicles   
2      10  UCBR8-60-B28hp2BmDPdntcQ              Music   
3      15  UCBR8-60-B28hp2BmDPdntcQ     Pets & Animals   
4      17  UCBR8-60-B28hp2BmDPdntcQ             Sports   

   item_snippet.assignable  
0                     True  
1                     True  
2                     True  
3                     True  
4                     True

# Pre-procesar los datos

## Añadir la información de categorías

Se debe mofificar la estructura del DataFrame cat (categories) ya que se deben extraer los atributos anidados en "items" y posteriormente, aquellos anidados en "snippet".

De esta forma se obtiene el DataFrame que tiene en cada fila la información de una categoría.

In [17]:
cat = pd.DataFrame(categories["items"].values.tolist())
cat["channelId"] = pd.DataFrame(cat["snippet"].values.tolist())["channelId"]
cat["title"] = pd.DataFrame(cat["snippet"].values.tolist())["title"]
cat["assignable"] = pd.DataFrame(cat["snippet"].values.tolist())["assignable"]
del cat["snippet"]

In [18]:
cat.head()

kind                                               etag  \
0  youtube#videoCategory  "ld9biNPKjAjgjV7EZ4EKeEGrhao/Xy1mB4_yLrHy_BmKm...   
1  youtube#videoCategory  "ld9biNPKjAjgjV7EZ4EKeEGrhao/UZ1oLIIz2dxIhO45Z...   
2  youtube#videoCategory  "ld9biNPKjAjgjV7EZ4EKeEGrhao/nqRIq97-xe5XRZTxb...   
3  youtube#videoCategory  "ld9biNPKjAjgjV7EZ4EKeEGrhao/HwXKamM1Q20q9BN-o...   
4  youtube#videoCategory  "ld9biNPKjAjgjV7EZ4EKeEGrhao/9GQMSRjrZdHeb1OEM...   

   id                 channelId             title  assignable  
0   1  UCBR8-60-B28hp2BmDPdntcQ  Film & Animation        True  
1   2  UCBR8-60-B28hp2BmDPdntcQ  Autos & Vehicles        True  
2  10  UCBR8-60-B28hp2BmDPdntcQ             Music        True  
3  15  UCBR8-60-B28hp2BmDPdntcQ    Pets & Animals        True  
4  17  UCBR8-60-B28hp2BmDPdntcQ            Sports        True

In [19]:
cat[["title", "assignable", "channelId", "etag"]].head()

title  assignable                 channelId  \
0  Film & Animation        True  UCBR8-60-B28hp2BmDPdntcQ   
1  Autos & Vehicles        True  UCBR8-60-B28hp2BmDPdntcQ   
2             Music        True  UCBR8-60-B28hp2BmDPdntcQ   
3    Pets & Animals        True  UCBR8-60-B28hp2BmDPdntcQ   
4            Sports        True  UCBR8-60-B28hp2BmDPdntcQ   

                                                etag  
0  "ld9biNPKjAjgjV7EZ4EKeEGrhao/Xy1mB4_yLrHy_BmKm...  
1  "ld9biNPKjAjgjV7EZ4EKeEGrhao/UZ1oLIIz2dxIhO45Z...  
2  "ld9biNPKjAjgjV7EZ4EKeEGrhao/nqRIq97-xe5XRZTxb...  
3  "ld9biNPKjAjgjV7EZ4EKeEGrhao/HwXKamM1Q20q9BN-o...  
4  "ld9biNPKjAjgjV7EZ4EKeEGrhao/9GQMSRjrZdHeb1OEM...

In [20]:
cat.index = cat["id"]
# cat.set_index("id")

In [21]:
cat[["title", "assignable", "channelId", "etag"]].head()

title  assignable                 channelId  \
id                                                           
1   Film & Animation        True  UCBR8-60-B28hp2BmDPdntcQ   
2   Autos & Vehicles        True  UCBR8-60-B28hp2BmDPdntcQ   
10             Music        True  UCBR8-60-B28hp2BmDPdntcQ   
15    Pets & Animals        True  UCBR8-60-B28hp2BmDPdntcQ   
17            Sports        True  UCBR8-60-B28hp2BmDPdntcQ   

                                                 etag  
id                                                     
1   "ld9biNPKjAjgjV7EZ4EKeEGrhao/Xy1mB4_yLrHy_BmKm...  
2   "ld9biNPKjAjgjV7EZ4EKeEGrhao/UZ1oLIIz2dxIhO45Z...  
10  "ld9biNPKjAjgjV7EZ4EKeEGrhao/nqRIq97-xe5XRZTxb...  
15  "ld9biNPKjAjgjV7EZ4EKeEGrhao/HwXKamM1Q20q9BN-o...  
17  "ld9biNPKjAjgjV7EZ4EKeEGrhao/9GQMSRjrZdHeb1OEM...

In [22]:
cat = cat[["title", "assignable", "channelId", "etag"]]

In [23]:
cat.index.names = ["category_id"]

In [24]:
cat.head()

title  assignable                 channelId  \
category_id                                                           
1            Film & Animation        True  UCBR8-60-B28hp2BmDPdntcQ   
2            Autos & Vehicles        True  UCBR8-60-B28hp2BmDPdntcQ   
10                      Music        True  UCBR8-60-B28hp2BmDPdntcQ   
15             Pets & Animals        True  UCBR8-60-B28hp2BmDPdntcQ   
17                     Sports        True  UCBR8-60-B28hp2BmDPdntcQ   

                                                          etag  
category_id                                                     
1            "ld9biNPKjAjgjV7EZ4EKeEGrhao/Xy1mB4_yLrHy_BmKm...  
2            "ld9biNPKjAjgjV7EZ4EKeEGrhao/UZ1oLIIz2dxIhO45Z...  
10           "ld9biNPKjAjgjV7EZ4EKeEGrhao/nqRIq97-xe5XRZTxb...  
15           "ld9biNPKjAjgjV7EZ4EKeEGrhao/HwXKamM1Q20q9BN-o...  
17           "ld9biNPKjAjgjV7EZ4EKeEGrhao/9GQMSRjrZdHeb1OEM...

### Verificar duplicados

In [25]:
cat["title"].nunique()

30

In [26]:
cat["title"].count()

31

In [27]:
cat["title"].unique()

array(['Film & Animation', 'Autos & Vehicles', 'Music', 'Pets & Animals',
       'Sports', 'Short Movies', 'Travel & Events', 'Gaming',
       'Videoblogging', 'People & Blogs', 'Comedy', 'Entertainment',
       'News & Politics', 'Howto & Style', 'Education',
       'Science & Technology', 'Movies', 'Anime/Animation',
       'Action/Adventure', 'Classics', 'Documentary', 'Drama', 'Family',
       'Foreign', 'Horror', 'Sci-Fi/Fantasy', 'Thriller', 'Shorts',
       'Shows', 'Trailers'], dtype=object)

In [28]:
cat["title"].value_counts()[cat["title"].value_counts() > 1]

Comedy    2
Name: title, dtype: int64

In [29]:
cat[cat["title"] == "Comedy"]["title"]

category_id
23    Comedy
34    Comedy
Name: title, dtype: object

### Duplicados: Conclusión
Los dos registros asociados a Comedy no tienen el mismo id, por lo que no es recomendable eliminar uno, ya que en el DataFrame principal pueden existir registros con cualquiera de los dos ids y en ambos casos deberán ser vinculados con la categoría "Comedy".

## Añadir el título de categoría al DF principal

In [30]:
cat_description = pd.DataFrame(cat["title"], index=cat.index)

In [31]:
cat_description.head()

title
category_id                  
1            Film & Animation
2            Autos & Vehicles
10                      Music
15             Pets & Animals
17                     Sports

In [32]:
cat_description.info()

<class 'pandas.core.frame.DataFrame'>
Index: 31 entries, 1 to 44
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   title   31 non-null     object
dtypes: object(1)
memory usage: 496.0+ bytes


In [33]:
cat_description.index = cat_description.index.astype(int)

In [34]:
df["category_id"] = df["category_id"].astype(int)

In [35]:
df = df.join(cat_description, how="left", rsuffix="_category", on="category_id")

In [36]:
# del df["title_category"]

In [37]:
df[["category_id", "title_category"]].head()

category_id title_category
0           10          Music
1           23         Comedy
2           23         Comedy
3           24  Entertainment
4           10          Music

## Formatear las fechas correctamente (**limpieza**)

In [38]:
df[["trending_date", "publish_time"]].head()

trending_date              publish_time
0      17.14.11  2017-11-10T17:00:03.000Z
1      17.14.11  2017-11-13T17:00:00.000Z
2      17.14.11  2017-11-12T19:05:24.000Z
3      17.14.11  2017-11-12T18:01:41.000Z
4      17.14.11  2017-11-09T11:04:14.000Z

In [39]:
df[["trending_date", "publish_time"]].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40881 entries, 0 to 40880
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   trending_date  40881 non-null  object
 1   publish_time   40881 non-null  object
dtypes: object(2)
memory usage: 638.9+ KB


Se identifican estos dos formatos divergentes:
* **trending_date**: YY.DD.MM
* **publish_time**: YYYY-MM-DD**T**00:00:00.000Z

Se modificarán ambos registros a un único formato: **DD-MM-YYYY**

También se identifica que las columnas almacenan datos de tipo **objeto**, que deberán ser modificados a tipo **datetime**.

### Tests (not in-place)

In [40]:
df["trending_date"].apply(lambda date: date[3:5]+"-"+date[-2:]+"-20"+date[:2]).head()

0    14-11-2017
1    14-11-2017
2    14-11-2017
3    14-11-2017
4    14-11-2017
Name: trending_date, dtype: object

In [41]:
df["publish_time"].apply(lambda date: date[8:10]+date[4:7]+"-"+date[:4]).head()

0    10-11-2017
1    13-11-2017
2    12-11-2017
3    12-11-2017
4    09-11-2017
Name: publish_time, dtype: object

In [42]:
# Se necesitará extraer también la hora de publicación.
df["publish_time"].apply(lambda datetime: datetime[11:-5]).head()

0    17:00:03
1    17:00:00
2    19:05:24
3    18:01:41
4    11:04:14
Name: publish_time, dtype: object

## Formateando la fecha

In [43]:
df["trending_date"] = df["trending_date"].apply(lambda date: date[3:5]+"-"+date[-2:]+"-20"+date[:2])

In [44]:
# Se necesitará extraer también la hora de publicación.
df = df.rename(columns={"publish_time":"publish_date"})
df["publish_time"] = df["publish_date"].apply(lambda datetime: datetime[11:-5])

In [45]:
df["publish_date"] = df["publish_date"].apply(lambda date: date[8:10]+date[4:7]+"-"+date[:4])

In [46]:
df[["trending_date", "publish_time", "publish_date"]].head()

trending_date publish_time publish_date
0    14-11-2017     17:00:03   10-11-2017
1    14-11-2017     17:00:00   13-11-2017
2    14-11-2017     19:05:24   12-11-2017
3    14-11-2017     18:01:41   12-11-2017
4    14-11-2017     11:04:14   09-11-2017

In [47]:
df[["trending_date", "publish_time", "publish_date"]].isna().sum()

trending_date    0
publish_time     0
publish_date     0
dtype: int64

## Modificando el tipo de dato

Al realizar la modificación del tipo de dato, el dtype **datetime** tiene un formato YY-MM-DD o YYYY-MM-DD, por lo que se cambiará a este formato (también válido según la consigna) al realizar la modificación de tipo de dato.

In [48]:
df["trending_date"] = pd.to_datetime(df["trending_date"], format="%d-%m-%Y")

In [49]:
df["publish_date"] = pd.to_datetime(df["publish_date"], format="%d-%m-%Y")

In [50]:
df[["trending_date", "publish_time", "publish_date"]].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40881 entries, 0 to 40880
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   trending_date  40881 non-null  datetime64[ns]
 1   publish_time   40881 non-null  object        
 2   publish_date   40881 non-null  datetime64[ns]
dtypes: datetime64[ns](2), object(1)
memory usage: 958.3+ KB


In [51]:
df["title_category"].isna().sum()

74

Existen registros para los cuales no se encontró la categoría asociada dentro del dataset de categorías. Por ello, se tendrá que tomar el valor de estas categorías del dataset de EE.UU.

# Tratado de datos faltantes

In [52]:
df.isna().sum()[df.isna().sum() > 0]

description       1296
title_category      74
dtype: int64

Tanto la descripción como el título de la categoría presentan valores faltantes. 

Se tratará cada caso individualmente.

## Datos faltantes de título de categoría

### Obtener la información

Se sabe que el documento .json de las categoría de EE.UU. debería contener la totalidad de categorías, por ello, se verificará si aquí se encuentra la(s) categoría(s) faltante(s).

In [53]:
url = "https://raw.githubusercontent.com/z3r0st/EB-2022-1-CC50/main/data/US_category_id.json"
categories_US = pd.read_json(url)

In [54]:
cat_US = pd.DataFrame(categories_US["items"].values.tolist())
cat_US["channelId"] = pd.DataFrame(cat_US["snippet"].values.tolist())["channelId"]
cat_US["title"] = pd.DataFrame(cat_US["snippet"].values.tolist())["title"]
cat_US["assignable"] = pd.DataFrame(cat_US["snippet"].values.tolist())["assignable"]
del cat_US["snippet"]

In [55]:
cat_US.head()

kind                                               etag  \
0  youtube#videoCategory  "m2yskBQFythfE4irbTIeOgYYfBU/Xy1mB4_yLrHy_BmKm...   
1  youtube#videoCategory  "m2yskBQFythfE4irbTIeOgYYfBU/UZ1oLIIz2dxIhO45Z...   
2  youtube#videoCategory  "m2yskBQFythfE4irbTIeOgYYfBU/nqRIq97-xe5XRZTxb...   
3  youtube#videoCategory  "m2yskBQFythfE4irbTIeOgYYfBU/HwXKamM1Q20q9BN-o...   
4  youtube#videoCategory  "m2yskBQFythfE4irbTIeOgYYfBU/9GQMSRjrZdHeb1OEM...   

   id                 channelId             title  assignable  
0   1  UCBR8-60-B28hp2BmDPdntcQ  Film & Animation        True  
1   2  UCBR8-60-B28hp2BmDPdntcQ  Autos & Vehicles        True  
2  10  UCBR8-60-B28hp2BmDPdntcQ             Music        True  
3  15  UCBR8-60-B28hp2BmDPdntcQ    Pets & Animals        True  
4  17  UCBR8-60-B28hp2BmDPdntcQ            Sports        True

In [56]:
cat_US["title"].nunique()

31

In [57]:
cat_US["title"].count()

32

Se identifica que el dataset de categorías de EE.UU. tiene 1 categoría más que el de Canadá. Se intuye que esta categoría es la correspondiente a los 74 registros faltantes.

### Comprobar que es la información faltante

In [58]:
for categorie in cat_US["title"].values.tolist():
    if categorie not in cat["title"].values.tolist():
        print(categorie)

Nonprofits & Activism


In [59]:
cat_US = cat_US.set_index("id")
cat_US.index = cat_US.index.astype(int)

In [60]:
cat_US[cat_US["title"] == "Nonprofits & Activism"]["title"]

id
29    Nonprofits & Activism
Name: title, dtype: object

In [61]:
df[df["title_category"].isna()]["category_id"].value_counts()

29    74
Name: category_id, dtype: int64

Aquí se ha comprobado que efectivamente el valor de título en los registros del DataFrame es "Nonprofits & Activism"

### Reemplazar los valores faltantes en el DataFrame

Ya que se ha verificado que se ha obtenido la información faltante de otra fuente relacionada, se puede reemplazar directamente con el valor hallado, sin necesidad de eliminar los registros ni aplicar una técnica de reemplazo (se están usando los valores reales, no unos simulados).

In [62]:
df["title_category"].fillna(value=cat_US.loc[29]["title"], inplace=True)

In [63]:
df["title_category"].isna().sum()

0

Se llenó exitosamente la información de todas las categorías.

## Datos faltantes en Descripción

In [64]:
df.columns

Index(['video_id', 'trending_date', 'title', 'channel_title', 'category_id',
       'publish_date', 'tags', 'views', 'likes', 'dislikes', 'comment_count',
       'thumbnail_link', 'comments_disabled', 'ratings_disabled',
       'video_error_or_removed', 'description', 'state', 'lat', 'lon',
       'geometry', 'title_category', 'publish_time'],
      dtype='object')

In [100]:
df[df["channel_title"] == df.iloc[0]["channel_title"]]["description"].head()

0       Eminem's new track Walk on Water ft. Beyoncé i...
214     Eminem's new track Walk on Water ft. Beyoncé i...
498     Eminem's new track Walk on Water ft. Beyoncé i...
791     Eminem's new track Walk on Water ft. Beyoncé i...
1226    Eminem's official 2017 SNL performance.\n \nWa...
Name: description, dtype: object

In [66]:
df[df["description"].isna()]["channel_title"].value_counts().head()

The View                        142
The Real Daytime                 49
Croton MEGA HIT 克頓傳媒2017爆款大劇     43
Motion Station                   39
Shalom Entertainment             38
Name: channel_title, dtype: int64

In [67]:
df[df["channel_title"] == "The View"]["channel_title"].count()

149

Para "The View", de 147 registros, 142 presentan descripción faltante.

Esto nos indica, por un lado, que sí existen registros con descripción para el canal, lo cual significa que sería posible llenar estos faltantes con cierto grado de precisión.

In [68]:
df[(df["description"].notna()) & (df["channel_title"] == "The View")]["description"]

470      The View co-hosts share their responses to tro...
18382    “This is a show where we respect one another a...
21427    The View co-hosts discuss whether the presiden...
22175    Clinton joined The View to discuss what she'd ...
28552             The word of the day in America is leaks.
28833             The word of the day in America is leaks.
33685    MORE FROM 'THE VIEW':Full episodes: http://abc...
Name: description, dtype: object

También se analizará las características de los registros del segundo canal con mayor cantidad de registros faltantes.

In [69]:
df[df["channel_title"] == "The Real Daytime"]["channel_title"].count()

133

Para "The Real Daytime", de 133 registros, 49 presentan descripción faltante.

Esto nos indica, por un lado, que existen bastantes registros con descripción para el canal, lo cual significa que sería posible llenar estos faltantes, pero es necesario comprobar si siempre se usa la misma descripción o esta varía.

In [70]:
df[(df["description"].notna()) & (df["channel_title"] == "The Real Daytime")]["description"].value_counts().head()

For how long could you be celibate?                                                     3
Do you think if a man truly loves his woman, another woman can take him away?           3
What do you think of Fergie's  version of the National Anthem?                          2
Do you think a child should automatically get the father's last name?                   2
If you had reservations before getting married, would you still walk down the aisle?    2
Name: description, dtype: int64

Como se puede visualizar, la moda tiene 3 apariciones, de un total de 84. De modo que no se presentan una descripción generalizada para el canal, es decir, la moda no es descriptiva del conjunto de descripciones. 

Por ello, se concluye como mejor estrategia, asignarle a las descripciones faltantes, el valor de una descripción del mismo canal escogida aleatoriamente.

In [85]:
import random

In [92]:
for channel in set(df[df["description"].isna()]["channel_title"].values.tolist()):
    n = len(df[(df["description"].notna()) & (df["channel_title"] == channel)])
    if n == 0: continue
    N = len(df[(df["description"].isna()) & (df["channel_title"] == channel)])
    
    new_values = []
    while(N > n):
        new_values += random.sample(df[(df["description"].notna()) & (df["channel_title"] == channel)]["description"].values.tolist(), k=n)
        N -= n
    new_values += random.sample(df[(df["description"].notna()) & (df["channel_title"] == channel)]["description"].values.tolist(), k=N)
    
    new_values = pd.Series(data=new_values, index=df[(df["description"].isna()) & (df["channel_title"] == channel)].index)
    for i in new_values.index:
        df.loc[i, "description"] = new_values.loc[i]
#     df[(df["description"].isna()) & (df["channel_title"] == channel)]["description"] = pd.Series(new_values)

In [101]:
df["description"].isna().sum()

722

In [99]:
channel_count, without_description_count = 0, 0
for channel in set(df[df["description"].isna()]["channel_title"].values.tolist()):
    channel_count += 1
    without_description_count += 1 > sum(df[df["channel_title"] == channel]["description"].notna())
print(f"In total, {channel_count} channels have videos registered without description, from which {without_description_count} haven't registered a description in any video")

In total, 303 channels have videos registered without description, from which 303 haven't registered a description in any video


Todos los valores faltantes son de canales que no publican descripciones para ninguno de sus videos, por ello se considera apropiado dejar ese valor como faltante.

No se ha decidido eliminar estos registros porque la descripción no es un dato importante para el conocimiento que buscamos obtener a partir de la exploración y procesamiento de los datos.

## Ordenar y organizar el DataFrame